sahi/utils/coco.py

```diff
@@ -1887,6 +1903,58 @@ def create_coco_dict(images, categories, ignore_negative_samples=False):
     return coco_dict
 
 
+def create_coco_prediction_array(images, ignore_negative_samples=False):
+    """
+    Creates COCO prediction array which is list of predictions
+
+    Arguments
+    ---------
+        images : List of CocoImage containing a list of CocoAnnotation
+        ignore_negative_samples : Bool
+            If True, images without predictions are ignored
+    Returns
+    -------
+        coco_prediction_array : List
+            COCO predictions array
+    """
+    predictions_array = []
+
+    num_images = len(images)
+    image_id = 1
```

Comment:
image ids should not be hard coded but fetched from the image objects as `image.id`. image ids dont have to be ordered.

Reply:
@PushpakBhoge  to fix this you can iterate over images instead of range()

Reply:
@fcakyon initially I thought of doing it in the similar way, but turns out `image_id` is not stored in instance of `CocoImage`. the `image_id` has been always `None` in those instances, I have verified this when I was working on some project. If we do it like this then image_id in json will be `null` that is invalid and will not work with `pycocotool`

I looked at function `sahi.utils.coco.create_coco_dict` which is associated with property `sahi.utils.Coco.json`. This function contain logic to generate Ground Truth coco json. This function work in the similar manner it hard code image_id and then increment it inside the loop. I am not aware why it is like this but I guess it might be required during coco slicing operation.

Hence I followed the similar structure and wrote logic in `sahi.utils.coco.create_coco_prediction_array`, trying not to modify existing flow. What would be your suggestion in this case ?

Reply:
@PushpakBhoge by default, id property is set as None, your are right on that.

However, when creating a CocoImage instance from a coco image json, id property is automatically set: https://github.com/obss/sahi/blob/79e28fecf48041b6c62fbca8e85b52ed71fa0375/sahi/utils/coco.py#L552

As a result, when instantiating a Coco object by dataset json path with this method: https://github.com/obss/sahi/blob/79e28fecf48041b6c62fbca8e85b52ed71fa0375/sahi/utils/coco.py#L964

all image CocoImage instances include valid `id`s

So you have to instantiate Coco instance as:

```python
from sahi.utils.coco import Coco
coco_obj = Coco.from_coco_dict_or_path("ground_truth.json")
```

Then all instances in `coco_obj.images` will include `id` property.

Reply:
`sahi.utils.coco.create_coco_dict` is designed to export a custom created coco dataset so it assigns the ids from scratch.

It is not the case when working with predictions (all images ids already include id)

Reply:
ok got what you are saying

we have "ground_truth.json" in coco format we load it in `coco_obj`, populate the predictions and then export predictions keeping image_ids in sync. This will work seamlessly if we added image ids from `image.id` 

Consider this we don't have annotations in coco format i.e. 'ground_truth.json', but we have annotations in some other format yolo, pascal or any else. Here to get coco metrics we need both predictions and ground truths in coco format. So in this case we will need to export both `json` and `predicion_array` properties right ?

Now, `json` property will assign the ids from scratch and `prediction_array` will look for ids in Image id in `CocoImage` which will be `None`

let's say we assigned ids at the time of creating `CocoImage` using `from_coco_image_dict`  method. In this case `CocoImage` object will have ids. `prediction_array` property will pick up those ids but `json` property will still reassign from scratch, won't that affect synch of image ids between the two json?

So, I think either we fetch id from `CocoImage` for both properties or we assign the ids from scratch for both properties. otherwise there might occur issues where the ids are not in sync.

for the first option we will have to make `<CocoImage_instance>.id` attribute as `NOT NULL`  that might have larger impact on other aspects of the framework. second option is simpler and have no impact on other functionality and both `json` and `prediction_array` can be thought of as temporary files to get coco-metrics. If anyone want to traceback to original ids a simple mapping using image name will be easy to create.

This was my thinking process to treat those two properties as temporary files, but if you think that we need to add ids from `CocoImage` will be happy to do that. let me know what are your thoughts.

P.S. - sorry for a very looooooooooooooong comment :bowtie:

Reply:
Thanks a lot @PushpakBhoge for your detailed explanations!

> Consider this we don't have annotations in coco format i.e. 'ground_truth.json', but we have annotations in some other format yolo, pascal or any else. Here to get coco metrics we need both predictions and ground truths in coco format. So in this case we will need to export both `json` and `predicion_array` properties right ?
> 
> Now, `json` property will assign the ids from scratch and `prediction_array` will look for ids in Image id in `CocoImage` which will be `None`

This is an awesome point!

Then we need to implement `create_coco_prediction_array` and `create_coco_dict` in a similar fashion that:
1. they iterate over CocoImage instances instead of integer range
2. when `image.id is None` then generate image_id's from a counter
3. when `image.id is not None` use `image.id` as the image_id

So `create_coco_dict` also requires an update (i can update it after this PR).

Reply:
okie great that sounds like a good idea

There is one thing that might go wrong, what if used added some images which have ids and some images which do not have ids ? This might create duplicate ids, incorrect id mapping etc.

This is off course invalid input scenario, but do we need to came up with a way to handle this. Maybe restricting the user in some way or is it fine if we mention the same thing in documentation and trust user to do right thing ?

Reply:
How about this:

when `create_coco_prediction_array ` or `create_coco_dict ` is called:
1. create a `image_id_set` using the id properties of the image instances (for the instances that contain id)
2. iterate over CocoImage instances instead of integer range
2.1. when image.id is None then generate image_id's from a counter, if generated id is in `image_id_set` increment again till generated number is not in `image_id_set`
2.2. when image.id is not None use image.id as the image_id

This way we can handle even if user forgets to assign id to some of the image instances.

Reply:
Sorry got busy

There is one case that might come. Let's we are assigning with counter and the counter reaches 100 then user send image with id 6, but that id is already assigned. This will create duplicate image_id which will make the coco invalid

I tried thinking of way how we can handle this but seems like there not much we can do without overengineering, hers's what I propose

1) We will recommend user in docs to add all `CocoImage` with existing image ids in incremental order first, then `CocoImage` with `image.id = None`
2) we will initialize a variable in `<Coco_Instance>.__init__` as `self.counter=1`, each time we receive image id if `image.id >= self.counter` we will set `self.counter = image_id + 1` othereise if `image_id < self.counter` we will throw exception saying `{image.id}  already taken, add CocoImage with predefined image_ids in incremental order first before adding CocoImage without image_ids to avoid this error` or something in line
3) in second stage when user add `CocoImage` with `image.id=None` we will internally increment the `self.counter` and set that in the `CocoImage` instance

so even if user added a duplicate id it won't go unnoticed and help keeing the exported the coco valid, since we are adding the ids during `<Coco_Instance>.add_image` user can quickly correct it then and there. Also techinically if image is a part of coco dataset then its better each image also have assigned id at the time of addition rather than waiting for it to export

granted that for some cases even if image_id is not present in the list it will throw exception
example `self.counter=1` and we added image with id=250 then we basically setting `self.counter=251` and it will basically not allow any image with id <=250. since the concept of primery_key (image_id) is that if assigned_id is 250 then there must images with ids upto 250 in original dataset. So this will serve the purpose better

so basically we will modify the add_image method like this
```python
     def add_image(self, image):
        """
        Adds image to this Coco instance

        Args:
            image: CocoImage
        """

        assert type(image) == CocoImage, "image must be a CocoImage instance"

        if image.id:
            if image.id >= self.counter:
                self.counter = image.id + 1
            else:
                raise Exception("{image.id}  already taken, add CocoImage with predefined image_ids in incremental order first before adding CocoImage without image_ids to avoid this error")
        else:
            image.id = self.counter
            self.counter+=1

        self.images.append(image)
```

Reply:
@PushpakBhoge you are right, we dont have to overengineer this.

I believe this might be simpler:

Add `image_id_setting: str` property to `Coco` instance init. Its value is `"auto"` by default. It can be set as "manual" by choice.

When `image_id_setting=="auto"`, then Image.id is ignored and image ids are generated automatically at export (as its the case now)

When `image_id_setting=="manual"`, then all image ids have to be manually set and `.add_image()` throws error when id is `None`

What do you think?

Reply:
@fcakyon  yes that sounds like perfect will make the necessary changes